File: engineering/concurrency/03-semaphore/skeleton/python/main.py
# 并发模型：threading
import threading
import time

class Semaphore:
    def __init__(self, n: int):
        # TODO: 初始化信号量，初始值为 n
        pass

    def acquire(self):
        # TODO: 如果计数器为 0，条件等待；否则减 1
        pass

    def release(self):
        # TODO: 加 1，如果有等待者，唤醒一个
        pass

def worker(sem, current, max_observed, mu, wid):
    sem.acquire()
    with mu:
        current[0] += 1
        if current[0] > max_observed[0]:
            max_observed[0] = current[0]
    time.sleep(0.01)
    with mu:
        current[0] -= 1
    sem.release()

def main():
    max_concurrent = 5
    total_workers = 50
    sem = Semaphore(max_concurrent)
    current = [0]
    max_observed = [0]
    mu = threading.Lock()
    threads = []

    for i in range(total_workers):
        t = threading.Thread(target=worker, args=(sem, current, max_observed, mu, i))
        threads.append(t)
        t.start()

    for t in threads:
        t.join()

    if max_observed[0] <= max_concurrent:
        print(f"PASS: max observed concurrency = {max_observed[0]} (limit = {max_concurrent})")
    else:
        print(f"FAIL: max observed concurrency = {max_observed[0]} (limit = {max_concurrent})")

if __name__ == "__main__":
    main()

File: engineering/concurrency/03-semaphore/skeleton/go/main.go
package main

import (
	"fmt"
	"sync"
	"sync/atomic"
	"time"
)

// Semaphore 计数信号量。
// TODO: 添加计数器和条件变量。
type Semaphore struct {
	mu sync.Mutex
	// TODO
}

func NewSemaphore(n int) *Semaphore {
	// TODO: 初始化信号量，初始值为 n
	return &Semaphore{}
}

func (s *Semaphore) Acquire() {
	s.mu.Lock()
	defer s.mu.Unlock()
	// TODO: 如果计数器为 0，条件等待；否则减 1
}

func (s *Semaphore) Release() {
	s.mu.Lock()
	defer s.mu.Unlock()
	// TODO: 加 1，如果有等待者，唤醒一个
}

func main() {
	const maxConcurrent = 5
	const totalWorkers = 50

	sem := NewSemaphore(maxConcurrent)
	var wg sync.WaitGroup
	var current int64
	var maxObserved int64

	for i := 0; i < totalWorkers; i++ {
		wg.Add(1)
		go func(id int) {
			defer wg.Done()
			sem.Acquire()

			c := atomic.AddInt64(&current, 1)
			for {
				m := atomic.LoadInt64(&maxObserved)
				if c <= m || atomic.CompareAndSwapInt64(&maxObserved, m, c) {
					break
				}
			}

			time.Sleep(10 * time.Millisecond)
			atomic.AddInt64(&current, -1)
			sem.Release()
		}(i)
	}

	wg.Wait()
	if maxObserved <= maxConcurrent {
		fmt.Printf("PASS: max observed concurrency = %d (limit = %d)\n", maxObserved, maxConcurrent)
	} else {
		fmt.Printf("FAIL: max observed concurrency = %d (limit = %d)\n", maxObserved, maxConcurrent)
	}
}
